### PR TITLE
With a single row GpuExplode tries to split the generator array

### LINKED ISSUE #10088: [BUG] GpuExplode single row split to fit cuDF limits
We have seen a case where a single row with strings (a 1MB sized string) and other columns, could have a list to `explode` by with many elements (10K elements for example). When we try to handle such an explode, we currently will not split the input, because it is a single row, but because of the repetition amount here we can go over cuDF column size limits.

The proposal is to at least do this in the `withRetry` case, where `splitInHalfByRows` for the explode case could have a special clause where for an input of 1 rows, it still can split the list we are exploding by, and replicate the row accordingly.

For example:
|col1|col2|
|-----|------|
|"very long string..."|[list of 10,000 items]|

Would become:
|col1|col2|
|-----|------|
|"very long string..."|[list of 5,000 items]|
|"very long string..."|[list of 5,000 items]|

or more rows, with less items in the list. We should be able to calculate the size of the row that would fit cuDF and then split the list accordingly and replicate the rest of the columns to go along.

### PR DESCRIPTION
**Pull request by abellina**
Closes #10088

The following PR adds special handling in `GpuGenerateExec` where a single row (cols + list to split) could be split in case of OOM.

The approach is very simple and leads to several retries, we simply split the column we are exploding (this would be a single list), in two halves. These are then attempted again, and following split and retry semantics, they can be split again until we are either successful, or the exploding column has 1 entry (we cannot split a single row with 1 array element).

### CODE REVIEW

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 1]
```diff
@@ -826,13 +827,99 @@ class GpuGenerateIterator(
...
+            // we are just going to split in half blindly until we can fit it.
+            val splits = withResource(explodedTbl) { _ =>
+              if (explodedTbl.getRowCount < 2) {
+                // we should never hit this
+                throw new GpuSplitAndRetryOOM(
+                  s"GPU OutOfMemory: cannot split a batch of 1 rows when " +
+                    s"the array has a single element!")
+              }
+              val splitIx = Seq(explodedTbl.getRowCount.toInt / 2)
+              explodedTbl.contiguousSplit(splitIx: _*)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:867]: Nit: Not sure why we're making a sequence-of-1 only to undo it later.
```suggestion
              val splitIx = explodedTbl.getRowCount.toInt / 2
              explodedTbl.contiguousSplit(splitIx)
```
or just:
```suggestion
              explodedTbl.contiguousSplit(explodedTbl.getRowCount.toInt / 2)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 2]
```diff
@@ -826,13 +827,99 @@ class GpuGenerateIterator(
...
+            }
+            val result = new Array[SpillableColumnarBatch](splits.size)
+            closeOnExcept(result) { _ =>
+              closeOnExcept(splits) { _ =>
+                (0 until splits.length).foreach { split =>
+                  val explodedTblToConvertBack = splits(split)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:873]:
```suggestion
                splits.zipWithIndex.foreach { case (explodedTblToConvertBack, split) =>
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 3]
```diff
@@ -826,13 +827,99 @@ class GpuGenerateIterator(
...
+                        cols(col) = if (col == generatorOffset) {
+                          withResource(
+                            colToReconstitute.reduce(
+                              ReductionAggregation.collectList(), DType.LIST)) { sc =>
+                            ColumnVector.fromScalar(sc, 1)
+                          }
+                        } else {
+                          withResource(tbl.getColumn(col).getScalarElement(0)) { sc =>
+                            ColumnVector.fromScalar(sc, 1)
+                          }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:888]: Seems like we're getting the first (and only) row of a column and then building a new column from that.  Why isn't this just incref of the original table column without copying anything?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 4]
```diff
@@ -826,13 +827,99 @@ class GpuGenerateIterator(
...
+                  val cols = new Array[ColumnVector](tbl.getNumberOfColumns)
+                  withResource(explodedTblToConvertBack) { _ =>
+                    val colToReconstitute = explodedTblToConvertBack.getTable.getColumn(0)
+                    closeOnExcept(cols) { _ =>
+                      (0 until tbl.getNumberOfColumns).foreach { col =>
+                        cols(col) = if (col == generatorOffset) {
+                          withResource(
+                            colToReconstitute.reduce(
+                              ReductionAggregation.collectList(), DType.LIST)) { sc =>
+                            ColumnVector.fromScalar(sc, 1)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:883]: This seems like a roundabout way to build a list column.  We already have the child column, we just need a simple offset vector of two rows, 0 and child column size, and then build a list column from that.

**Review by jlowe — COMMENTED** [review]
Discussed this with @revans2 and he mentioned uses like posexplode might cause issues since the positions would need fixup after this mid-row split logic.

**Comment by abellina** [comment]
Thanks for the review @jlowe and @revans2. I confirmed this approach does not work for `posexplode` => the position column is added early. Good news is that we'll have a test to cover it.

**Comment by kuhushukla** [comment]
One other consideration would be how this change performs wrt to the CPU and what(if) we need to change any recommendations to get close to or better than the CPU perf.

**Comment by abellina** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 5]
```diff
@@ -367,7 +371,7 @@ abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGene
     val inputRows = inputBatch.numRows()
 
     // if the number of input rows is 1 or less, cannot split
-    if (inputRows <= 1) return Array()
+    //if (inputRows <= 1) return Array()
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:374]: I would love to see us push `inputSplitIndices` into the individual generate expressions. But that can be follow on as it really is just tech debt.

Also why is the row commented out?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 6]
```diff
@@ -528,13 +532,35 @@ abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGene
...
+      fixUpOffset: Long): Table = {
     if (position) {
-      if (outer) {
+      val toFixUp = if (outer) {
         inputTable.explodeOuterPosition(genOffset)
       } else {
         inputTable.explodePosition(genOffset)
       }
+      // the position column is at genOffset and the exploded column at genOffset + 1
+      val posColToFix = toFixUp.getColumn(genOffset)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:547]: nit toFixUp needs to either be under withResource, or at least a closeOnException.

**Review by revans2 — COMMENTED** [review]
In general it looks good.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala (hunk 5 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:374]: Nice catch, that should not be commented out.

I'll look into the inputSplitIndices refactor.

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
Note  that the integration tests don't exercise the split path because it isn't setup correctly. I am going to punt on split integration tests. I do have unit tests in this PR that exercise the path. 

Filed this as follow on work: https://github.com/NVIDIA/spark-rapids/issues/10185

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala (hunk 5 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala]: Filed this for the refactor of `inputSplitIndices` and more... https://github.com/NVIDIA/spark-rapids/issues/10187. I don't want to block this issue further.

**Comment by abellina** [comment]
@revans2 @jlowe this is ready for another look.

@kuhushukla the approach taken here is purely on retry and we are recovering from the cuDF column size limit exception. In order to make this perform better, we can change our size calculation to account for the cuDF column size limit up front. Lets add this as part of a refactor of the generator split code. That code only knows how to split by rows right now and that was a bigger change, so we decided to do the retry way instead. If we can split the input not just by rows, but within rows (before we retry), we can generalize some of this code and avoid several retry attempts.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 7]
```diff
@@ -528,13 +532,35 @@ abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGene
...
+        }
+        withResource(fixedUpPosCol) { _ =>
+          val newCols = new Array[ColumnVector](toFixUp.getNumberOfColumns)
+          (0 until genOffset).foreach { b =>
+            newCols(b) = toFixUp.getColumn(b)
+          }
+          newCols(genOffset) = fixedUpPosCol
+          (genOffset + 1 until toFixUp.getNumberOfColumns).foreach { a =>
+            newCols(a) = toFixUp.getColumn(a)
+          }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:560]:
```suggestion
          val newCols = (0 until toFixUp.getNumberOfColumns).map { i =>
            if (i == genOffset) {
              fixedUpPosCol
            } else {
              toFixUp.getColumn(i)
          }.toArray
```

**Review by jlowe — APPROVED** [review]
Minor code cleanup suggestion,  otherwise lgtm.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala (hunk 7 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:560]: Let me make that change, that's better.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala]: Thanks @jlowe this is handled in https://github.com/NVIDIA/spark-rapids/pull/10131/commits/123b5d87708040c3bc296e6b5372f617ec66a7d4

**Comment by abellina** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Pull request merged**